### PR TITLE
Gh-521: Document Gremlin integration in the REST API

### LINKED ISSUE #521: Document Gremlin in Gaffer REST API for 2.3

### PR DESCRIPTION
**Pull request by tb06904**
Documents the new way to use Gremlin from within the REST API along with the new explain endpoints too. Moves the custom features to their own page and started off the user auth page.

# Related issue

- Resolve #521

### CODE REVIEW

File: docs/administration-guide/gaffer-deployment/gremlin.md [hunk 1]
```diff
@@ -1,15 +1,22 @@
-# Connecting Gremlin to Gaffer
+# Using Gremlin in Gaffer
 
 It is possible to use Gremlin as an alternative querying language in Gaffer. To
 make Gremlin available however, there are some additional steps that need to be
-taken to connect to a Gaffer graph via this interface.
+taken ensure it is setup correctly.
```
Comment (cn337131) [inline docs/administration-guide/gaffer-deployment/gremlin.md]:
```suggestion
taken to ensure it is setup correctly.
```

File: docs/administration-guide/gaffer-deployment/gremlin.md [hunk 2]
```diff
@@ -19,193 +26,66 @@ using the following dependency definition:
...
 
-### Configuring the Gafferpop Library
+## Configuring the Gafferpop Library
 
 The `gafferpop.properties`, file is the configuration for the Gaffer
-implementation of Tinkerpop (a.k.a Gafferpop). Most of the set up here is for
-the construction of the Gafferpop Graph instance which we want to make run with
-the `store.properties` we've already configured. An example `gaffer.properties`
-would look like the following:
+implementation of Tinkerpop (a.k.a Gafferpop). If using the REST API there is no
```
Comment (cn337131) [inline docs/administration-guide/gaffer-deployment/gremlin.md]: Not sure we need this repeated, could just be GafferPop as you have already explained what it is above.

File: docs/administration-guide/gaffer-deployment/gremlin.md [hunk 3]
```diff
@@ -19,193 +26,66 @@ using the following dependency definition:
...
+mandatory properties you need to set since you already will have configured the
+Graph in the existing `store.properties` file. However, adding some default
+values in for operation modifiers such as a limit for `GetAllElement`
```
Comment (cn337131) [inline docs/administration-guide/gaffer-deployment/gremlin.md]:
```suggestion
values in for operation modifiers, such as a limit for `GetAllElement`
```

File: docs/administration-guide/gaffer-deployment/gremlin.md [hunk 4]
```diff
@@ -19,193 +26,66 @@ using the following dependency definition:
...
+operations is good practise.
```
Comment (cn337131) [inline docs/administration-guide/gaffer-deployment/gremlin.md]:
```suggestion
operations, is good practise.
```

File: docs/administration-guide/security/user-control.md [hunk 5]
```diff
@@ -1,3 +1,49 @@
...
+
 !!! info "Work in Progress"
 
-    This page is under construction.
\ No newline at end of file
+    This page is under construction.
+
+The security layer for Gaffer is currently only enforced by the REST API. We
+recommend restricting users such that they do not have access to the cluster and
+by only allowing users to connect to Gaffer via the REST API.
```
Comment (cn337131) [inline docs/administration-guide/security/user-control.md]: Not sure on this sentence phrasing, the bit after and could do with being clearer.

File: docs/administration-guide/security/user-control.md [hunk 6]
```diff
@@ -1,3 +1,49 @@
...
+
+In the REST API the `User` object is constructed via a [`UserFactory`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/rest/factory/UserFactory.html).
+Under the Spring REST API this has an abstract implementation class
```
Comment (cn337131) [inline docs/administration-guide/security/user-control.md]:
```suggestion
In the Spring REST API an abstract implementation of this class is used,
```

File: docs/administration-guide/security/user-control.md [hunk 7]
```diff
@@ -1,3 +1,49 @@
...
+`AbstractUserFactory` to allow passing HTTP headers to be used for
```
Comment (cn337131) [inline docs/administration-guide/security/user-control.md]:
```suggestion
`AbstractUserFactory`, which is then used in the passing of HTTP headers for
```

Think these comments make it a bit clearer what you are trying to say here. Feel free to re-write however though

File: docs/administration-guide/security/user-control.md [hunk 8]
```diff
@@ -1,3 +1,49 @@
...
+authentication.
+
+Currently, we only provide one implementation of this, the `UnknownUserFactory`
+which will always just return `new User()`. The `UnknownUserFactory` is the default
+factory, to specify the user factory class define the
+`gaffer.user.factory.class` [REST property](../gaffer-config/config.md#application-properties).
```
Comment (cn337131) [inline docs/administration-guide/security/user-control.md]: This paragraph reads a little weird. Maybe just simplify to something like:
```
Currently, there is a single default implementation this; the `UnknownUserFactory` which simply returns a `new User()`. 
To specify the user factory class define the`gaffer.user.factory.class` [REST property](../gaffer-config/config.md#application-properties).
```

File: docs/administration-guide/security/user-control.md [hunk 9]
```diff
@@ -1,3 +1,49 @@
...
+
+## Writing a User Factory
+
+To authenticate your users, you will need to extend the `AbstractUserFactory` class
```
Comment (cn337131) [inline docs/administration-guide/security/user-control.md]:
```suggestion
To authenticate your users you will need to extend the `AbstractUserFactory` class
```

File: docs/administration-guide/security/user-control.md [hunk 10]
```diff
@@ -1,3 +1,50 @@
...
+REST API.
+
+In the REST API the `User` object is constructed via a [`UserFactory`](https://gchq.github.io/Gaffer/uk/gov/gchq/gaffer/rest/factory/UserFactory.html).
+In the Spring REST API an abstract implementation of this class is used,
+`AbstractUserFactory`, which is then used in the passing of HTTP headers for
+authentication.
+
+Currently, there is a single default implementation of this; the
+`UnknownUserFactory` which simply returns a new `User` with `UNKNOWN` as the
+user ID. The `UnknownUserFactory` is the default factory, to specify the user
```
Comment (cn337131) [inline docs/administration-guide/security/user-control.md]:
```suggestion
user ID. To specify the user
```
Not sure you need to repeat that UnknownUserFactory is the default as you have already said that in the previous sentence

File: docs/administration-guide/gaffer-deployment/gremlin.md [hunk 11]
```diff
@@ -19,193 +26,65 @@ using the following dependency definition:
...
-The `gafferpop.properties`, file is the configuration for the Gaffer
-implementation of Tinkerpop (a.k.a Gafferpop). Most of the set up here is for
-the construction of the Gafferpop Graph instance which we want to make run with
-the `store.properties` we've already configured. An example `gaffer.properties`
-would look like the following:
+The `gafferpop.properties`, file is the configuration for GafferPop. If using
+the REST API there is no mandatory properties you need to set since you already
+will have configured the Graph in the existing `store.properties` file. However,
+adding some default values in for operation modifiers, such as a limit for
+`GetAllElement` operations, is good practise.
```
Comment (wb36499) [inline docs/administration-guide/gaffer-deployment/gremlin.md:66]: Practice

File: docs/administration-guide/gaffer-deployment/gremlin.md [hunk 12]
```diff
@@ -19,193 +26,65 @@ using the following dependency definition:
...
-# This class is necessary for correctly forwarding the user to Gaffer
-authorization: {
-  authorizer: uk.gov.gchq.gaffer.tinkerpop.server.auth.GafferPopAuthoriser
-}
-```
-
-The `authorizer` should always be the `GafferPopAuthoriser` as this is what
-handles denying invalid queries for GafferPop and passing the user on to the
-Gaffer graph for fine grained security.
+A full breakdown of of the available properties is as follows:
```
Comment (wb36499) [inline docs/administration-guide/gaffer-deployment/gremlin.md:74]: Double 'of'

File: docs/administration-guide/gaffer-deployment/gremlin.md (hunk 11 above)
Comment (wb36499) [inline docs/administration-guide/gaffer-deployment/gremlin.md]:
```suggestion
`GetAllElement` operations, is good practice.
```

File: docs/administration-guide/gaffer-deployment/gremlin.md (hunk 12 above)
Comment (wb36499) [inline docs/administration-guide/gaffer-deployment/gremlin.md]:
```suggestion
A full breakdown of the available properties is as follows:
```

**Review by wb36499 — APPROVED** [review]

**Review by cn337131 — APPROVED** [review]

**Pull request merged**